strawberryfields/engine.py

```diff
@@ -531,161 +544,541 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
+        """
+        response = self._get(
+            "/jobs/{}/result".format(job_id), {"Accept": "application/x-numpy"},
+        )
+        if response.status_code == 200:
+            # Read the numpy binary data in the payload into memory
+            with io.BytesIO() as buf:
+                buf.write(response.content)
+                buf.seek(0)
+                samples = np.load(buf)
+            return Result(samples, is_stateful=False)
+        raise RequestFailedError(self._format_error_message(response))
+
+    def cancel_job(self, job_id: str):
+        """Cancels a job.
 
-        Raises:
-            Exception: In case a job could not be submitted or completed.
-            TypeError: In case a job is already queued and a user is trying to submit a new job.
+        Args:
+            job_id (str): the job ID
         """
-        if self.jobs:
-            raise OneJobAtATimeError(
-                "A job is already queued. Please reset the engine and try again."
-            )
+        response = self._patch(
+            "/jobs/{}".format(job_id), data={"status", JobStatus.CANCELLED.value}
+        )
+        if response.status_code == 204:
+            return
+        raise RequestFailedError(self._format_error_message(response))
 
-        kwargs.update(program.run_options)
-        job_content = self._get_blackbird(program=program, **kwargs).serialize()
-        job = self._queue_job(job_content)
+    def ping(self) -> bool:
+        """Tests the connection to the remote backend.
 
-        try:
-            while not job.is_failed and not job.is_complete:
-                job.reload()
-                time.sleep(self.polling_delay_seconds)
-        except KeyboardInterrupt:
-            if job.id:
-                print("Job {} is queued in the background.".format(job.id.value))
-            else:
-                self.reset()
-                raise JobNotQueuedError("Job was not sent to server. Please try again.")
+        Returns:
+            bool: ``True`` if the connection is successful, and ``False`` otherwise
+        """
+        response = self._get("/healthz")
+        return response.status_code == 200
+
+    def _get(
+        self, path: str, headers: Dict[str, str] = None, **kwargs
+    ) -> requests.Response:
+        return self._request(RequestMethod.GET, path, headers, **kwargs)
+
+    def _post(
+        self, path: str, headers: Dict[str, str] = None, **kwargs
+    ) -> requests.Response:
+        return self._request(RequestMethod.POST, path, headers, **kwargs)
+
+    def _patch(
+        self, path: str, headers: Dict[str, str] = None, **kwargs
+    ) -> requests.Response:
+        return self._request(RequestMethod.PATCH, path, headers, **kwargs)
+
+    def _request(
+        self, method: RequestMethod, path: str, headers: Dict[str, str] = None, **kwargs
+    ) -> requests.Response:
+        headers = {} if headers is None else headers
+        return getattr(requests, method.value)(
```

Comment:
It might be worth having a small auxiliary naming here for readabilitiy such as
```python
request_method = getattr(requests, method.value)
return request_method(urljoin(self.base_url, path),
            headers={"Authorization": self.token, **headers},
            **kwargs
        )
```

Reply:
Thanks for pointing that out! Updated.

Reply:
Hmm, after going through `_get`, `_post`, `_patch` and `_request`, I'm wondering, is there any advantage in having `_request` and having a `RequestMethod` enum class?

Maybe the explicit values of `get`, `post`, and `patch` could simply be used in their corresponding methods to making the return statement for example for `_get`:
```python
return requests.get(urljoin(self.base_url, path), headers={"Authorization": self.token, **headers}, **kwargs)
```

But let me know if I missed something or you'd rather keep the implementation as is, for future extendability, this is more of a suggestion after parsing each of them separately. :)

Reply:
Yes, that's a perfectly valid point :+1:

I was on the fence myself about whether that introduced additional indirection for no substantial benefit. My thinking was to avoid repeating shared elements, like the `"Authorization"` header, between requests; but now that you mention it, it seems that is the _only_ thing that is actually duplicated, so it's probably better to just be explicit and direct about it.

I will look at simplifying the code as you described. Thanks!

Reply:
I've removed the wrapper methods for sending requests, and the `create_job`, `get_job` (etc.) methods now call the relevant `requests` method directly.